File: python/src/codecs.cc
// Copyright 2020-2022 PANDA GmbH

#include "codecs.h"

#include <wavelet_buffer/codecs/image_codec.h>
#include <wavelet_buffer/codecs/jpeg_codecs.h>

#include <pybind11/numpy.h>
#include <pybind11/operators.h>

#include "blaze_utils.h"


void WrapCodecAlgorithms(py::module* m) {
  using drift::SignalN2D;
  using drift::img::codecs::ColorSpace;
  using drift::img::codecs::GrayJpegCodec;
  using drift::img::codecs::HslJpegCodec;
  using drift::img::codecs::IImageCodec;
  using drift::img::codecs::RgbJpegCodec;

  auto module = m->def_submodule("codecs", "Codecs");

  module.doc() =
      "Codecs to encode and decode pictures to\\from numpy images."
      "Numpy image has planar layout: shape = [channel_number, height, width]";

  auto base = py::class_<IImageCodec>(module, "Base");

  auto decode = [](IImageCodec& self, const py::bytes& data) {
    SignalN2D image;
    if (!self.Decode(data, &image, 0)) {
      throw py::value_error("Failed to decode data");
    }

    return SignalN2DToPyArray(image);
  };

  auto encode = [](IImageCodec& self, const py::array& data,
                   size_t start_channel) {
    auto shape = data.shape();
    std::vector signal_shape = {static_cast<size_t>(shape[1]),
                                static_cast<size_t>(shape[2])};
    std::string blob;
    if (!self.Encode(PyArrayToSignalN2D(data), &blob, start_channel)) {
      throw py::value_error("Failed to encode data");
    }
    return py::bytes(blob);
  };

  base.def("decode", decode, "Decodes pictures from string to a numpy image",
           py::arg("data"));

  base.def("encode", encode,
           "Encodes a numpy image to string in some image format",
           py::arg("data"), py::arg("start_channel") = 0);

  base.def("channel_number",
           [](IImageCodec& self) { return self.channel_number(); });

  py::class_<RgbJpegCodec, IImageCodec>(module, "RgbJpeg")
      .def(py::init(
               [](float write_quality) { return RgbJpegCodec(write_quality); }),
           py::arg("write_quality") = 1.f);

  py::class_<HslJpegCodec, IImageCodec>(module, "HslJpeg")
      .def(py::init(
               [](float write_quality) { return HslJpegCodec(write_quality); }),
           py::arg("write_quality") = 1.f);

  py::class_<GrayJpegCodec, IImageCodec>(module, "GrayJpeg")
      .def(py::init([](float write_quality) {
             return GrayJpegCodec(write_quality);
           }),
           py::arg("write_quality") = 1.f);
}

File: python/src/wavelet_buffer/img/codecs.py
"""Collections of image codecs"""
import wavelet_buffer._wavelet_buffer as impl


class BaseCodec:
    """Interface for codes to decode/encode images from/to different image formats"""

    def decode(self, blob: bytes) -> "np.ndarray[np.float32]":
        """Decode an image from bytes to a numpy image
        Args:
            blob: serialized image e.g. JPEG
        Returns:
            image as numpy array
        """
        raise NotImplemented()

    def encode(self, data: "np.ndarray[np.float32]", start_channel: int = 0) -> bytes:
        """Encode a numpy image to byte string
        Args:
            data: numpy image
            start_channel: start channel for conversion
        Returns:
            serialized image e.g. JPEG
        """
        raise NotImplemented()

    def channel_number(self) -> int:
        """Return number ofr channel RGB=3, Gray=1"""
        raise NotImplemented()


class RgbJpeg(BaseCodec):
    """Codec for JPEG <-> numpy RGB image conversion"""

    def __init__(self, write_quality: float = 1.0):
        """
        Args:
            write_quality: quality of encoded image. Should be from 0 to 1.0
        """
        self._impl = impl.RgbJpeg(write_quality)

    def decode(self, blob: bytes) -> "np.ndarray[np.float32]":
        return self._impl.decode(blob)

    def encode(self, data: "np.ndarray[np.float32]", start_channel: int = 0) -> bytes:
        return self._impl.encode(data, start_channel)

    def channel_number(self) -> int:
        return self._impl.channel_number()


class HslJpeg(BaseCodec):
    """Codec for JPEG <-> numpy HSL image conversion"""

    def __init__(self, write_quality: float = 1.0):
        """
        Args:
            write_quality: quality of encoded image. Should be from 0 to 1.0
        """
        self._impl = impl.HslJpeg(write_quality)

    def decode(self, blob: bytes) -> "np.ndarray[np.float32]":
        return self._impl.decode(blob)

    def encode(self, data: "np.ndarray[np.float32]", start_channel: int = 0) -> bytes:
        return self._impl.encode(data, start_channel)

    def channel_number(self) -> int:
        return self._impl.channel_number()


class GrayJpeg(BaseCodec):
    """Codec for JPEG  <-> numpy monochrome image conversion"""

    def __init__(self, write_quality: float = 1.0):
        """
        Args:
            write_quality: quality of encoded image. Should be from 0 to 1.0
        """
        self._impl = impl.GrayJpeg(write_quality)

    def decode(self, blob: bytes) -> "np.ndarray[np.float32]":
        return self._impl.decode(blob)

    def encode(self, data: "np.ndarray[np.float32]", start_channel: int = 0) -> bytes:
        return self._impl.encode(data, start_channel)

    def channel_number(self) -> int:
        return self._impl.channel_number()
